streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+        line = media_playlist.readline()
+    
+    # Set the features we need to access easily while performing the concatenation.
+    # Features like codec, channel_layout, resolution, etc... .
+    self._set_features(streams_map)
+  
+  def _set_features(self, streams_map: Dict[str, OutputStream]) -> None:
+    """Get the audio and video codecs and other relavent stream features
+    from the matching OutputStream in the `streams_map`, this will be used
+    in the codec matching process in the concat_xxx() methods, but the codecs
+    that will be written in the final concatenateded master playlist will be
+    the codecs from stream_info dictionary(in HLS syntax).
+    """
+    
+    # We can't depend on the stream_info['CODECS'] to get the codecs from because
+    # this is only present for STREAM-INF, this makes it harder to get codecs for
+    # audio segments. Also if we try to get the audio codecs from one of the 
+    # #EXT-X-STREAM-INF tags we will have to match these codecs with each stream,
+    # for example, a codec attribute might be CODECS="avc1,ac-3,mp4a,opus",
+    # the video codec is put first then the audio codecs are put in 
+    # lexicographical order(by observation), which isn't nesseccary the same order of 
+    # #EXT-X-MEDIA in the master playlist, thus there is no solid baseground for 
+    # matching the codecs using the information in the master playlist.
+    
+    output_stream: Optional[OutputStream] = None
+    lines = self.content.split('\n')
+    for i in range(len(lines)):
+      # Don't use #EXT-X-MAP to get the codec, because HLS specs says
+      # it is an optional tag.
+      line = lines[i]
+      if line.startswith('#EXTINF'):
+        line = lines[i + 1]
+        if line.startswith('#EXT-X-BYTERANGE'):
+          line = lines[i + 2]
+        file_name = os.path.basename(line)
+        # Index the file_name and don't use dict.get() .
+        # There MUST be a match.
+        output_stream = streams_map[file_name]
+        break
+    
+    assert output_stream, 'No media file found in this media playlist'
+    self.codec = output_stream.codec
+    if isinstance(output_stream, VideoOutputStream):
+      self.resolution = output_stream.resolution
+    elif isinstance(output_stream, AudioOutputStream):
+      self.channel_layout = output_stream.layout
+      # We will get the language from the stream_info because the stream information
+      # is provided by Packager.  We might have mixed 3-letter and 2-letter format
+      # from the Streamer, but Packager reduces them all to 2-letter language tag.
+      self.language = _unquote(self.stream_info.get('LANGUAGE', '"und"'))
+    elif isinstance(output_stream, TextOutputStream):
+      self.language = _unquote(self.stream_info.get('LANGUAGE', '"und"'))
+  
+  def write(self, dir_name: str) -> None:
+    file_path = os.path.join(dir_name, _unquote(self.stream_info['URI']))
+    with open(file_path, 'w') as media_playlist:
```

Comment:
There are other places where "media_playlist" is a MediaPlaylist instance.  Here, it's a file instance.  That makes it confusing when searching the code for media_playlist.write(), because they mean two different things in these two contexts.  Could you please rename this file instance to something else?